component/encryptor/encryptor.go

```diff
@@ -0,0 +1,368 @@
+	return internal.EComponentPriority.LevelMid()
+}
+
+func (e *Encryptor) Configure(isParent bool) error {
+	log.Trace("Encryptor::Configure :  %s", e.Name())
+
+	conf := encryptorOptions{}
+	err := config.UnmarshalKey(e.Name(), &conf)
+	if err != nil {
+		log.Err("Encryptor::Configure : config error [invalid config attributes]")
+		return fmt.Errorf("error reading config for %s: %w", e.Name(), err)
+	}
+
+	key := os.Getenv("ENCRYPTION_KEY")
+	if key == "" {
+		key = conf.EncryptionKey
+		if key == "" {
+			log.Err("Encryptor::Configure : encryption key not set")
+			return fmt.Errorf("encryption key not set")
+		}
+	}
+
+	e.encryptionKey, err = base64.StdEncoding.DecodeString(key)
+	if err != nil {
+		log.Err("Encryptor::Configure : error decoding encryption key")
+		return fmt.Errorf("error decoding encryption key: %w", err)
+	}
+
+	e.blockSize = 1024 * 1024 // default block size is 1MB
+	if config.IsSet(e.Name()+".block-size-mb") && conf.BlockSize > 0 {
+		e.blockSize = conf.BlockSize * 1024 * 1024
+	}
+
+	e.mountPointCipher = "/mnt/cipher/"
+	if config.IsSet(e.Name() + ".encrypted-mount-path") {
+		e.mountPointCipher = conf.EncryptedMountPath
+	}
+
+	e.lastChunkMeta = &LastChunkMeta{
+		farthestBlockSeen: -1,
+		paddingLength:     0,
+	}
+
+	log.Info("Encryptor::Configure : block size %d, encrypted mount path %s", e.blockSize, e.mountPointCipher)
+	return nil
+}
+
+func (e *Encryptor) Start(ctx context.Context) error {
+	return nil
+}
+func (e *Encryptor) CommitData(opt internal.CommitDataOptions) error {
+
+	defer e.handle.Close()
+	e.lastChunkMeta.Lock()
+	defer e.lastChunkMeta.Unlock()
+
+	if e.lastChunkMeta.farthestBlockSeen != -1 {
+		paddingLengthByte := make([]byte, 8)
+		binary.BigEndian.PutUint64(paddingLengthByte, uint64(e.lastChunkMeta.paddingLength))
+		endoffset := (e.lastChunkMeta.farthestBlockSeen + 1) * (int64(e.blockSize) + MetaSize)
+		n, err := e.handle.WriteAt(paddingLengthByte, endoffset)
+		log.Info("Encryptor::CommitData : writing %d bytes, padding length %d at offset %d", n, e.lastChunkMeta.paddingLength, endoffset)
+		if err != nil {
+			log.Err("Encryptor: Error writing padding length to file: %s", err.Error())
+			return err
+		}
+		e.lastChunkMeta.farthestBlockSeen = -1
+		e.lastChunkMeta.paddingLength = 0
+	}
+	return nil
+}
+
+func (e *Encryptor) CreateFile(options internal.CreateFileOptions) (*handlemap.Handle, error) {
+	log.Info("Encryptor::createFile : %s", e.mountPointCipher+options.Name)
+
+	handle := handlemap.NewHandle(options.Name)
+	if handle == nil {
+		log.Trace("Encryptor::createFile : Failed to create handle for file: %s", options.Name)
+		return nil, syscall.EFAULT
+	}
+	handle.Mtime = time.Now()
+
+	fileHandle, err := os.OpenFile(e.mountPointCipher+options.Name, os.O_RDWR|os.O_CREATE, options.Mode)
+	if err != nil {
+		log.Trace("Encryptor::createFile : Error creating file: %s", err.Error())
+		return nil, err
+	}
+
+	e.handle = fileHandle
```

Comment:
>e.handle = fileHandle [](http://example.com/codeflow?start=1&length=21)

You will need a map to maintain these file handles, right ? By saving it as an instance, you are assuming that the file handle will be closed before another one is opened, otherwise data will be corrupt.

Reply:
Yep will handle for parallel files in a separate PR.